zio-http/jvm/src/main/scala/zio/http/netty/client/NettyClientDriver.scala

```diff
@@ -74,23 +87,28 @@ final case class NettyClientDriver private[netty] (
       }
       .map { jReq =>
         val closeListener: GenericFutureListener[ChannelFuture] = { (_: ChannelFuture) =>
-          // If onComplete was already set, it means another fiber is already in the process of fulfilling the promises
-          // so we don't need to fulfill `onResponse`
           nettyRuntime.unsafeRunSync {
-            onComplete.interrupt && onResponse.fail(NettyClientDriver.PrematureChannelClosure)
+            // wait for ClientFailureHandler.exceptionCaught to complete onFailure.
+            onFailure.await.flatMap { error =>
+              // If onComplete was already set, it means another fiber is already in the process of fulfilling the promises
+              // so we don't need to fulfill `onResponse`
+              onComplete.interrupt && onResponse.fail(error)
+            }.fork
```

Comment:
I have a really bad feeling about this. We're relying on `onFailure` being always completed, otherwise we're risking memory leaks. Can we ensure that there are no cases where this callback is invoked but `onFailure` was never completed? Perhaps it's better to use `poll` instead and avoid `fork` altogether? In cases that it returns `None`, we can fallback to the error in within the `ChannelFuture` (if present) and finally a generic error as previously if that's not set either.

AFAIK Netty invokes everything within the same channel in a single thread, and the `exceptionCaught` method of the handlers should be invoked before the channel future is closed.